jetty-servlet/src/test/java/org/eclipse/jetty/servlet/ServletContextHandlerTest.java

```diff
@@ -884,6 +884,38 @@ public void onStartup(Set<Class<?>> c, ServletContext ctx) throws ServletExcepti
         assertThat("Response", response, containsString("Hello World"));
     }
 
+    @Test
+    public void testAddJspFile() throws Exception
+    {
+        ContextHandlerCollection contexts = new ContextHandlerCollection();
+        _server.setHandler(contexts);
+
+        ServletContextHandler root = new ServletContextHandler(contexts, "/");
+        ServletHolder jspServlet = new ServletHolder();
+        jspServlet.setName("jsp");
+        jspServlet.setHeldClass(FakeJspServlet.class);
+        root.addServlet(jspServlet, "*.jsp");
+        class JSPAddingSCI implements ServletContainerInitializer
+        {
+            @Override
+            public void onStartup(Set<Class<?>> c, ServletContext ctx) throws ServletException
+            {
+                try
+                {
+                    ServletRegistration rego = ctx.addJspFile("some.jsp", "/path/to/some.jsp");
+                    rego.addMapping("/somejsp/*");
+                }
+                catch (Exception e)
+                {
+                    fail(e);
+                }
+            }
+        }
+        
+        root.addBean(new MySCIStarter(root.getServletContext(), new JSPAddingSCI()), true);
+        _server.start();
+    }
```

Comment:
All other tests in this class have code that actually make a HTTP request to test the functionality.
This new test only adds the dynamic JSP but does not test that making a HTTP request to it works.
Is that intended?

Reply:
Also, should we not have 3 tests for this functionality? No mapping, partial mapping, existing mapping?

Reply:
I can't add a test for a jsp file because jetty-servlet cannot depend on apache-jsp module. I've added an assertion that tests that the mapped jsp servlet is the one matched by the ServletHandler for the given path, that's as much as we can do.

I also added tests for an existing full mapping and an existing preliminary mapping.

Reply:
Can we add a jspFile via this API in the test-spec webapp as an integration test?

Reply:
That was already part of the initial commit: the test-spec-webapp calls addJsp via a ServletContainerInitiaizer.  See https://github.com/eclipse/jetty.project/blob/jetty-10.0.x-4433-addJspFile/tests/test-webapps/test-servlet-spec/test-container-initializer/src/main/java/com/acme/initializer/FooInitializer.java#L108 and https://github.com/eclipse/jetty.project/blob/jetty-10.0.x-4433-addJspFile/tests/test-webapps/test-servlet-spec/test-spec-webapp/src/main/java/com/acme/test/AnnotationTest.java#L255

Reply:
great!  can we then check that the test-distribution will see any fail for this?

Reply:
Done.